packages/tool-cache/src/tool-cache.ts

```diff
@@ -454,6 +455,42 @@ export function findAllVersions(toolName: string, arch?: string): string[] {
   return versions
 }
 
+// versions-manifest
+//
+// typical pattern of a setup-* action that supports JIT would be:
+// 1. resolve semver against local cache
+//
+// 2. if no match, download
+//   a. query versions manifest to match
+//   b. if no match, fall back to source if exists (tool distribution)
+//   c. with download url, download, install and preprent path
+
+export type IToolRelease = mm.IToolRelease
+export type IToolReleaseFile = mm.IToolReleaseFile
+
+export async function getManifestFromUrl(url: string): Promise<IToolRelease[]> {
+  const http: httpm.HttpClient = new httpm.HttpClient('tool-cache')
+
+  //let versions: IToolRelease[] =
+  return (await http.getJson<IToolRelease[]>(url)).result || []
+  //return
+}
+
+export async function findFromManifest(
+  versionSpec: string,
+  stable: boolean,
+  manifest: IToolRelease[]
+): Promise<IToolRelease | undefined> {
+  // wrap the internal impl
+  const match: mm.IToolRelease | undefined = await mm._findMatch(
+    versionSpec,
+    stable,
+    manifest
+  )
+
+  return match
+}
+
```

Comment:
I think that function `installAsset` is a good candidate for toolkit:
```
export async function installReleaseFile(release: IToolRelease): Promise<void> {
  const releaseFileUrl = release.files[0];
  // 1. download file
  // 2. unpack archive
  // 3. run setup.ps1 script from archive if it exists
}
```
This logic will work for all tools so it can be reused between actions.
`setup.ps1` script copies archive content to hosted tool-cache directory and and perform any other actions if needs.
Please see [nix example](https://github.com/actions/python/blob/master/installers/nix-setup-template.ps1) and [win example](https://github.com/actions/python/blob/master/installers/win-setup-template.ps1).

Reply:
Actions are pure javascript and the runner carries the runtime ( a version of node 12 to run them).  The runner currently runs actions inside of containers (from the host) and self-hosted machines that do not have powershell.  We even run actions inside alpine containers (8 MB) and the runner carries an alpine version of node to do that.

Hold off on this and it will be clearer when we integrate these changes into something like setup-node with progressive checks on cache, dist and then source.

Reply:
Also note that there's already existing methods tool-cache to install into the cache. See it used here: https://github.com/actions/setup-node/blob/master/src/installer.ts#L173  As part of this change, I don't want to revisit the download, extract and cache.  We're just covering another way to discover an url by version spec.